### ISSUE TITLE
can't 'jbuilder install' automatically imply 'jbuilder build @install

### ISSUE THREAD

**Issue by UnixJunkie** [janestreet/jbuilder#215]

**Comment by UnixJunkie** [comment]
Of course, when I order to install things, things should be build before.

    $ jbuilder install
    The following <package>.install are missing:
    [...]
    You need to run: jbuilder build @install

**Comment by diml** [comment]
The rationale for `jbuilder install` not building things is that sometimes people build as their user and install as root. It's not uncommon to do: `make && sudo make install`. In this case, we don't want the build to run as root

**Comment by UnixJunkie** [comment]
In that case, maybe an option to the install command would do the trick.
For example:

    jbuilder install --build-if-necessary / jbuilder install --build

Also, let's assume that jbuilder does nothing when things are already build.
Your use case would still work:

    jbuilder build @install && sudo jbuilder install

**Comment by rgrinberg** [comment]
`alias 'jbuilder install --build-if-necessary'='jbuilder build && jbuilder install'` should solve the problem personally for you.

Personally, I don't think such an option is a good idea.

**Comment by UnixJunkie** [comment]
I think that in most makefiles, make install triggers the build if necessary.

**Comment by rgrinberg** [comment]
I suppose we could  make it more makefile like by making uninstall/install aliases. But then you can't really pass extra params to install. And you always want to avoid buildings things with sudo as diml mentioned. The benefits seem quite low...

**Comment by UnixJunkie** [comment]
If other users want/expect this, let's hope they will manifest themselves here.
If I am the only one bothered, it is fine to close this after some time.

**Comment by dra27** [comment]
I personally think the present behaviour is better, but the GNU standards, for example, recommend that `make install` should build the program if `make all` hasn't been run (with the caveat that if `make all` has been previously run, then `make install` should not alter the build tree so that `make all; sudo make install` shouldn't build anything as root). See [their standards document](https://www.gnu.org/prep/standards/html_node/Standard-Targets.html) for more info.

**Comment by copy** [comment]
I'm also in favor installing implying building. This is the behavior I would expect, and the opposite can lead installing stale files when one forgets to run build. It would also be consistent with the behavior of `jbuilder exec` after #152 is implemented.

**Comment by rgrinberg** [comment]
@copy you're referring to the enhancement that `$ jbuilder exec foo` should rebuild `foo` first?

**Comment by copy** [comment]
> @copy you're referring to the enhancement that $ jbuilder exec foo should rebuild foo first?

Yes.

**Comment by dra27** [comment]
As a compromise between the security consideration of not wanting to build as root but wanting the convenience, perhaps we could emit a warning if the build artefacts are stale. Or possibly emit the warning if the artefacts are stale and the various detections for sudo/su/root reveal that jbuilder may be being run elevated, but otherwise rebuild first? These would be hard to do for make but relatively easy to do for jbuilder.
